src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,173 @@
+            }
+
+            return;
+        }
+
+        if (!_projectSet.Contains(projectKey))
+        {
+            var configurationFilePath = FilePathNormalizer.Normalize(projectInfo.SerializedFilePath);
+            _logger.LogInformation("Found no existing project key for project key '{0}'. Assuming new project.", projectKey.Id);
+
+            AddProject(configurationFilePath, projectInfo);
+            return;
+        }
+
+        _logger.LogInformation("Project info changed for project '{0}'", projectKey.Id);
+
+        EnqueueUpdateProject(projectKey, projectInfo);
+    }
+
+    private void AddProject(string configurationFilePath, RazorProjectInfo projectInfo)
+    {
+        var projectFilePath = FilePathNormalizer.Normalize(projectInfo.FilePath);
+        var intermediateOutputPath = Path.GetDirectoryName(configurationFilePath).AssumeNotNull();
+        var rootNamespace = projectInfo.RootNamespace;
+
+        var projectKey = _projectService.AddProject(projectFilePath, intermediateOutputPath, projectInfo.Configuration, rootNamespace, projectInfo.DisplayName);
+        _projectSet.Add(projectKey);
+
+        _logger.LogInformation("Project configuration file added for project '{0}': '{1}'", projectFilePath, configurationFilePath);
+        EnqueueUpdateProject(projectKey, projectInfo);
+    }
+
+    private void UpdateProject(ProjectKey projectKey, RazorProjectInfo? projectInfo)
+    {
+        if (projectInfo is null)
+        {
+            ResetProject(projectKey);
+            return;
+        }
+
+        _logger.LogInformation("Actually updating {project} with a real projectInfo", projectKey);
+
+        var projectWorkspaceState = projectInfo.ProjectWorkspaceState ?? ProjectWorkspaceState.Default;
+        var documents = projectInfo.Documents;
+        _projectService.UpdateProject(
+            projectKey,
+            projectInfo.Configuration,
+            projectInfo.RootNamespace,
+            projectInfo.DisplayName,
+            projectWorkspaceState,
+            documents);
+    }
+
+    private async Task UpdateAfterDelayAsync(ProjectKey projectKey)
+    {
+        await Task.Delay(EnqueueDelay).ConfigureAwait(true);
+
+        var delayedProjectInfo = ProjectInfoMap[projectKey];
+        UpdateProject(projectKey, delayedProjectInfo.ProjectInfo);
+    }
+
+    private void EnqueueUpdateProject(ProjectKey projectKey, RazorProjectInfo? projectInfo)
+    {
+        if (!ProjectInfoMap.ContainsKey(projectKey))
+        {
+            ProjectInfoMap[projectKey] = new DelayedProjectInfo();
+        }
+
+        var delayedProjectInfo = ProjectInfoMap[projectKey];
+        delayedProjectInfo.ProjectInfo = projectInfo;
+
+        if (delayedProjectInfo.ProjectUpdateTask is null || delayedProjectInfo.ProjectUpdateTask.IsCompleted)
+        {
+            delayedProjectInfo.ProjectUpdateTask = UpdateAfterDelayAsync(projectKey);
```

Comment:
If batching stays, consider using the `AsyncBatchingWorkQueue` rather than rolling something custom

Reply:
I'm game to try it, though this was just copied from the ProjectConfigurationStateSynchronizer. See my comment on batching above.

Reply:
Switched to AsyncBatchingWorkQueue in https://github.com/dotnet/razor/pull/10182/commits/0586df0ebd84dfc375067a18dc6b9acd9aca1b93